### PR TITLE
Enable TSA with ramping constraints

### PR DESCRIPTION
**Pull request by julia1071**
RR constraint can now be used when modeling with typical days

### CODE REVIEW

File: src/components/technologies/genericTechnologies/conv1.py [hunk 1]
```diff
@@ -838,25 +839,73 @@ def bind_disjunctions(dis, t):
             )
 
         else:
+            if data["config"]["optimization"]["typicaldays"]["N"]["value"] == 0:
+                input_aux = self.input
```
Comment (JeanWi) [inline src/components/technologies/genericTechnologies/conv1.py:843]: could we refactor this var into input_aux_rr or something similiar to not confuse it with input_aux in the technology class?

File: src/components/technologies/genericTechnologies/conv1.py [hunk 2]
```diff
@@ -838,25 +839,73 @@ def bind_disjunctions(dis, t):
...
+                ):
+                    sequence = data["k_means_specs"]["sequence"]
+                elif (
+                    data["config"]["optimization"]["typicaldays"]["method"]["value"]
+                    == 2
+                ):
+                    sequence = self.sequence
+
+                # init bounds at full res
+                bounds_RR_full = {
```
Comment (JeanWi) [inline src/components/technologies/genericTechnologies/conv1.py:858]: please keep variable names lower case (i.e. bounds_rr_full). I know it doesnt look super beautiful, but thats just what it is... :D

Comment (JeanWi) [reply src/components/technologies/genericTechnologies/conv1.py:858]: (holds for everything else as well where we have a RR)

File: src/components/technologies/genericTechnologies/conv1.py [hunk 3]
```diff
@@ -838,25 +839,73 @@ def bind_disjunctions(dis, t):
             )
 
         else:
+            if data["config"]["optimization"]["typicaldays"]["N"]["value"] == 0:
+                input_aux = self.input
+                set_t = self.set_t_performance
```
Comment (JeanWi) [inline src/components/technologies/genericTechnologies/conv1.py:844]: also this one to set_t_rr?

File: src/components/technologies/genericTechnologies/stor.py [hunk 4]
```diff
@@ -653,8 +654,12 @@ def init_ramping_down_rate_operation_in(const):
                             # -rampingRate <= input[t] - input[t-1]
                             return (
                                 -ramping_rate
-                                <= self.input[t, self.main_car]
-                                - self.input[t - 1, self.main_car]
+                                <= self.input[
```
Comment (JeanWi) [inline src/components/technologies/genericTechnologies/stor.py:657]: The if statement above (l 642) is different from the other technologies - correct?

File: src/energyhub.py [hunk 5]
```diff
@@ -108,13 +108,22 @@ def _perform_preprocessing_checks(self):
                 for node in topology["nodes"]:
                     for tec_name in self.data.technology_data[period][node]:
                         tec = self.data.technology_data[period][node][tec_name]
-                        if ("ramping_time" in tec.processed_coeff.dynamics) and (
-                            tec.processed_coeff.dynamics["ramping_time"] != -1
+                        if (
```
Comment (JeanWi) [inline src/energyhub.py:111]: This looks not correct to me:
- if TSA is used or not is not determined by method but by N>0.
- the exception is raised stating method 2, but it is raised for both 1 and 2

**Review by JeanWi — CHANGES_REQUESTED** [review]
As far as I see it we still need to do the rr + tsa with the following tecs:
- hydro open
- gas turbine
- heat pump

File: src/components/technologies/genericTechnologies/conv1.py (hunk 1 above)
Comment (julia1071) [reply src/components/technologies/genericTechnologies/conv1.py:843]: done

File: src/components/technologies/genericTechnologies/conv1.py (hunk 2 above)
Comment (julia1071) [reply src/components/technologies/genericTechnologies/conv1.py:858]: done

File: src/components/technologies/genericTechnologies/conv1.py (hunk 3 above)
Comment (julia1071) [reply src/components/technologies/genericTechnologies/conv1.py:844]: done

File: src/components/technologies/genericTechnologies/stor.py (hunk 4 above)
Comment (julia1071) [reply src/components/technologies/genericTechnologies/stor.py:659]: yes because storage has no performance type 1

File: src/energyhub.py (hunk 5 above)
Comment (julia1071) [reply src/energyhub.py:111]: yes you are right, its fixed now!

**Comment by julia1071** [comment]
> As far as I see it we still need to do the rr + tsa with the following tecs:
> 
> * hydro open
> * gas turbine
> * heat pump

Only generic technologies have dynamics and ramping constraints because it can be very case specific how you want to implement it.

**Review by JeanWi — APPROVED** [review]

**Review by lucabert01 — APPROVED** [review]

**Pull request merged**